python/mxnet/gluon/data/dataloader.py

```diff
@@ -655,3 +655,149 @@ def __del__(self):
+                    data = zip(*data)
+                    return [default_batchify_fn(i) for i in data]
+                else:
+                    data = np.asarray(data)
+                    return nd.array(data, dtype=data.dtype)
+
+    num_workers : int, default 0
+        The number of multiprocessing workers to use for data preprocessing.
+    pin_memory : boolean, default False
+        If ``True``, the dataloader will copy NDArrays into pinned memory
+        before returning them. Copying from CPU pinned memory to GPU is faster
+        than from normal CPU memory.
+    pin_device_id : int, default 0
+        The device id to use for allocating pinned memory if pin_memory is ``True``
+    prefetch : int, default is `num_workers * 2`
+        The number of prefetching batches only works if `num_workers` > 0.
+        If `prefetch` > 0, it allow worker process to prefetch certain batches before
+        acquiring data from iterators.
+        Note that using large prefetching batch will provide smoother bootstrapping performance,
+        but will consume more shared_memory. Using smaller number may forfeit the purpose of using
+        multiple worker processes, try reduce `num_workers` in this case.
+        By default it defaults to `num_workers * 2`.
+    thread_pool : bool, default False
+        If ``True``, use threading pool instead of multiprocessing pool. Using threadpool
+        can avoid shared memory usage. If `DataLoader` is more IO bounded or GIL is not a killing
+        problem, threadpool version may achieve better performance than multiprocessing.
+    timeout : int, default is 120
+        The timeout in seconds for each worker to fetch a batch data. Only modify this number
+        unless you are experiencing timeout and you know it's due to slow data loading.
+        Sometimes full `shared_memory` will cause all workers to hang and causes timeout. In these
+        cases please reduce `num_workers` or increase system `shared_memory` size instead.
+    """
+    def __init__(self, dataset, batch_size=None, shuffle=False, sampler=None,
+                 last_batch=None, batch_sampler=None, batchify_fn=None,
+                 num_workers=0, pin_memory=False, pin_device_id=0,
+                 prefetch=None, thread_pool=False, timeout=120):
+        super(PrefetchedDataLoader, self).\
+            __init__(dataset, batch_size, shuffle, sampler,
+                     last_batch, batch_sampler, batchify_fn,
+                     num_workers, pin_memory, pin_device_id,
+                     prefetch, thread_pool, timeout)
+        self.refresh()
+
+    def __iter__(self):
+        if self._iter is None:
+            self.refresh()
+        t = self._iter
+        self._iter = None # ensure a single iter would not using twice.
+        for item in t:
+            yield item
+        if self._iter is None: # ensure we do not waste any exist iter by mistake
+            self.refresh()
+
+    def refresh(self):
```

Comment:
I suppose `refresh` is not intended as public function? use `_refresh` if so.

Reply:
In my opinion and your old comment, some user will modify the training data after training process is end, in such case, a manual refresh is needed.
Anyway, I merge the auto_reload flag into dataloader, and set it to true. The reason is that in mxnet 2.0, dataloader in nopython mode prefetch data, and since the `self._dataset` is a private value, it should not be modified, thus the prefetch will not bring any breaking changes.

Reply:
Why is changing the behavior of MXNet 1.x to match 2.x not a breaking change for 1.x?